Rewrite publishing intro (#16916)
dotnet/docs: docs/core/deploying/index.md

> ---

Reviewer: The switches are a little weird and change across different SDK versions. I didn't see a link in the body here to that.

Author: I'm not sure what exactly you're referencing. I have a switch table in the intro and these are the commands to produce those types of apps. .NET Core 2.2 introduced that kind of odd bridge between 2.1 and 3.0 but didn't have the defaults 3.0 has. Since 2.2 is out of support and not in this table, it's a lot easier to describe.

title: Application publishing
description: Learn about the ways to publish a .NET Core application. You can publish either a self-contained app or a runtime-dependent app.
ms.date: 01/31/2020
---
# .NET Core application publishing overview

> Applications you create with .NET Core can be published in two different ways. Each way affects how a user installs and runs your app. Publishing your app self-contained includes the .NET Core runtime and libraries along with all app dependencies and produces a platform-specific executable. Publishing your app runtime-dependent only contains your app with it's dependencies, and produces a cross-platform binary. Starting with .NET Core SDK 3.0, publishing an app runtime-dependent also produces a platform-specific executable.

Reviewer 2: ```suggestion
Applications you create with .NET Core can be published in two different ways. Publishing your app self-contained produces *.dll* files that contain your app and its dependencies, the .NET Core runtime and libraries, and a platform-specific executable. Publishing your app runtime-dependent produces a cross-platform *.dll* file that contains your app and its dependencies. Starting with .NET Core SDK 3.0, publishing an app runtime-dependent also produces a platform-specific executable.
```

Could also briefly summarize the difference of needing or not needing the runtime to be installed on the target machine.

If you prefer not to accept the suggestion, note that "it's" should still be corrected to "its".

Author: Whoops, I totally swapped that it's... 🐕 

I like it. How about this adjustment though? I feel like it's important to state that this really affects how a user runs the app. Do they or don't they need to have the runtime installed. (further detailed in the article sections)

```suggestion
Applications you create with .NET Core can be published in two different ways, each way affects how a user runs your app. Publishing your app self-contained produces a *.dll* file containing your app and includes: its dependencies, the .NET Core runtime and libraries, and a platform-specific executable. Publishing your app runtime-dependent produces a cross-platform *.dll* file that contains your app and its dependencies. Starting with .NET Core SDK 3.0, publishing an app runtime-dependent also produces a platform-specific executable for your current platform.
```

Reviewer 2: Looks good, but I would make the first two sentences separate sentences or link them in some other way than just a comma.

> Applications you create with .NET Core can be published in two different modes, and depending on the mode, affects how a user runs your app. Publishing your app self-contained produces a *.dll* file containing your app and includes: its dependencies, the .NET Core runtime and libraries, and a platform-specific executable. Publishing your app runtime-dependent produces a cross-platform *.dll* file that contains your app and its dependencies. Starting with .NET Core SDK 3.0, publishing an app runtime-dependent also produces a platform-specific executable for your current platform.

Reviewer: Maybe: 
```suggestion
Applications you create with .NET Core can be published in two different modes, and depending on the mode, affects how a user runs your app. Publishing your app self-contained produces an application that contains the .NET Core runtime and libraries, in addition to your application and its dependencies. Publishing your app runtime-dependent produces an application that contains only your application and its dependencies, and users of your application will need to separately install the .NET Core runtime. Both types of applications include a platform specific executable by default. Runtime-dependent applications without an executable can be created and these applications are cross-platform. 
```

The following table outlines the basic commands used to publish an app as runtime-dependent or self-contained, per SDK:

| Type                                                          | SDK 3.1 | SDK 3.0 | SDK 2.1 | Command | 
| ------------------------------------------------------------- | --------| ------- | ------- | ------- |
| [runtime-dependent cross-platform binary](#runtime-dependent) | ✔️       | ✔️       | ✔️       | [`dotnet publish`](../tools/dotnet-publish.md) |
| [runtime-dependent executable](#runtime-dependent)            | ✔️       | ✔️       |          | [`dotnet publish`](../tools/dotnet-publish.md) |
| [self-contained executable](#self-contained)                  | ✔️       | ✔️       | ✔️       | [`dotnet publish -r <RID>`](../tools/dotnet-publish.md) |

## Runtime-dependent

> Apps published as runtime-dependent are cross-platform and don't include the .NET Core runtime. However, the user of your app is required to install the .NET Core runtime.

Author: The runtime-dependent app contains a platform-specific executable **and** a cross-platform dll for 3.0. With 2.1 only the dll is included. Regardless, both SDKs produce the dll that is cross-platform.

A *dll* file is created when you publish your app. For example, if you have an app named **word_reader**, a file named *word_reader.dll* is created. Apps published in this way are run with the `dotnet <filename.dll>` command and can be run on any platform.

> Along with the *dll* file containing your app, a platform-specific executable can be published if the `-r <RID>` parameter is used. The *dll* is still created and remains cross-platform while the executable isn't. Continuing the example above, if you publish the **word_reader** app and target Windows, a *word_reader.exe* executable is created along with *word_reader.dll*. When targeting Linux or macOS, a *word_reader* executable is created along with *word_reader.dll*. For more information about RIDs, see [.NET Core RID Catalog](../rid-catalog.md).

Reviewer: Suggest removing "if the ...is used" here. Then after this and the next paragraph use samples or a little chart about what combination of switches you need - it's not just -r, but also --self-contained false. And then maybe do the "continuing the example" after the samples or table.

Starting with .NET Core 3.0 SDK, an executable is always created for your current platform when you omit the `-r` parameter.

If the app uses a NuGet package that has platform-specific implementations, dependencies for all platforms are copied to the publish folder along with the app.

> A runtime-dependent app can be created with the [`dotnet publish`](../tools/dotnet-publish.md) command.

Reviewer: "for the RID of the current machine"

```dotnet
dotnet publish
```

A runtime-dependent app can be created for a specific platform, for example, Linux 64-bit.

```dotnet
dotnet publish -r linux-x64 --self-contained false
```

### Advantages

- **Small deployment**\
Only your app and its dependencies are distributed. The .NET Core runtime and libraries are installed by the user and all apps share the runtime.

- **Cross-platform**\
Your app and any .NET-based library runs on other operating systems. You don't need to define a target platform for your app. For more information about the .NET file format, see [.NET Assembly File Format](../../standard/assembly/file-format.md).

- **Uses the latest patched runtime**\

> The app uses the latest runtime (within the targeted major-minor family of .NET Core) installed on the target system. This means your app automatically uses the latest patched version of the .NET Core runtime. This default behavior can be overridden.

Reviewer: Can we link to version roll forward here.

### Disadvantages

- **Requires pre-installing the runtime**\
Your app can run only if the version of .NET Core your app targets is already installed on the host system. You can configure roll-forward behavior for the app to either require a specific version of .NET Core or allow a newer version of .NET Core. For more information, see [runtime-dependent apps roll forward](../versions/selection.md#framework-dependent-apps-roll-forward).

- **.NET Core may change**\

> It's possible for the .NET Core runtime and libraries to be updated on the machine where the app is run. In rare cases, this may change the behavior of your app if you use the .NET Core libraries, which most apps do.

Reviewer: ```suggestion
It's possible for the .NET Core runtime and libraries to be updated on the machine where the app is run. In rare cases, this may change the behavior of your app if you use the .NET Core libraries, which most apps do. Use runtime roll forward [link] to control what runtimes your application can use. 
```

> - **Per-platform executables**\

Reviewer: This isn't a disadvantage of runtime-dependent as this is also true of self contained applications.

Author: I think changing the item to **Use `dotnet` to run your app** better communicates the point. Then I'll describe how with 3.0 you can generate per-platform executables while remaining not-self-contained.

*.NET Core 3.0 and later versions*\
If you want your users to start your app directly and avoid the `dotnet <filename.dll>` command, you must publish your app for every platform.

## Self-contained

Publishing your app as self-contained produces a platform-specific executable. The output publishing folder contains all components of the app, including the .NET Core libraries and target runtime. The app is isolated from other .NET Core apps and doesn't use a locally installed shared runtime. The user of your app isn't required to download and install .NET Core.

> The executable binary produced for the specified target platform. For example, if you have an app named **word_reader**, and you publish a self-contained executable for Windows, a *word_reader.exe* file is created. Publishing for Linux or macOS, a *word_reader* file is created. The target platform and architecture is specified with the `-r <RID>` parameter. For more information about RIDs, see [.NET Core RID Catalog](../rid-catalog.md).

Reviewer: ```suggestion
The executable binary is produced for the specified target platform. For example, if you have an app named **word_reader**, and you publish a self-contained executable for Windows, a *word_reader.exe* file is created. Publishing for Linux or macOS, a *word_reader* file is created. The target platform and architecture is specified with the `-r <RID>` parameter. For more information about RIDs, see [.NET Core RID Catalog](../rid-catalog.md).
```

If the app has platform-specific dependencies, such as a NuGet package containing platform-specific dependencies, these are copied to the publish folder along with the app.

A self-contained executable can be created with the [`dotnet publish -r`](../tools/dotnet-publish.md) command.

```dotnet

> dotnet publish -r <RID>

Reviewer: Above you give an example RID which I think is nicer.

```

### Advantages

- **Control .NET Core version**\
You control which version of .NET Core is deployed with your app.

- **Platform-specific targeting**\
Because you have to publish your app for each platform, you know where your app will run. If .NET Core introduces a new platform, users can't run your app on that platform until you release a version targeting that platform. You can test your app for compatibility problems before your users run your app on the new platform.

### Disadvantages

- **Larger deployments**\

> Because your app includes the .NET Core runtime and all of your app dependencies, the hard drive space required is greater than a [runtime-dependent](#runtime-dependent) version.

Reviewer: ```suggestion
Because your app includes the .NET Core runtime and all of your app dependencies, the download size and hard drive space required is greater than a [runtime-dependent](#runtime-dependent) version.
```

  > [!TIP]
  > You can reduce the size of your deployment on Linux systems by approximately 28 MB by using .NET Core [*globalization invariant mode*](https://github.com/dotnet/runtime/blob/master/docs/design/features/globalization-invariant-mode.md). This forces your app to treat all cultures like the [invariant culture](xref:System.Globalization.CultureInfo.InvariantCulture?displayProperty=nameWithType).

- **Control .NET Core version**\

> .NET Core (distributed with your app) can only be upgraded by releasing a new version of your app.

Reviewer: ```suggestion
.NET Core Runtime (distributed with your app) can only be upgraded by releasing a new version of your app. You are responsible for supplying an updated version of your application for security patches to the .NET Core Runtime. 
```

## See also

- [Deploying .NET Core Apps with .NET Core CLI.](deploy-with-cli.md)
- [Deploying .NET Core Apps with Visual Studio.](deploy-with-vs.md)

> - [Packages, Metapackages, and Frameworks.](../packages.md)

Reviewer: @dsplaisted Can you check this link https://docs.microsoft.com/en-us/dotnet/core/packages. It appears out of date. If so, let's remove it here.

Reviewer 3: Yes, that is super out of date.  I think we should delete that page entirely, as it doesn't look like any of it applies any more. @mairaw @KathleenDollard

Author: @dsplaisted None of it applies anymore? What about for 2.1? I'll remove the link but I think we should open another issue to evaluate this.

---
title: Application publishing

> description: Learn about the ways to publish a .NET Core application. You can publish either a self-contained app or a runtime-dependent app.

Author: I don't think it is. When you publish you do have the same code binary apart from same setting/runtime-specific files/dependencies etc. It does reconfigure the runtime config files published with your app based on the publish mode.

It's always been described in this way: when you publish, the entire output you get **is your app** and it's considered self-contained or runtime-dependent.

01. A platform-specific executable.

    Depending on the commands used, platform-specific executables can be published [self-contained](#self-contained) or [runtime-dependent](#runtime-dependent).

    Commands that produce an executable:

    | Type                                    | SDK Version   | Command |
    | --------------------------------------- | ------------- | ------- |
    | [runtime-dependent](#runtime-dependent) | 3.1, 3.0      | [`dotnet publish`](../tools/dotnet-publish.md) |
    |                                         | 2.1           | Not supported. |
    | [self-contained](#self-contained)       | 3.1, 3.0, 2.1 | [`dotnet publish -r <RID>`](../tools/dotnet-publish.md) |

    In general, platform-specific apps can be created for platforms other than your current platform. The `-r <RID>` switch sets the target [runtime-identifier](../rid-catalog.md).

01. A runtime-dependent cross-platform *dll* file.

>     Commands that produce a runtime-dependent *dll* app:

Author: I think it's better to call this an "app" as it's not a library which you would assume with the .dll extension. It has an entry point and is meant to be run. I'm not totally sure how to articulate this though, it's a confusing file extension to be chosen...

Thoughts?

>     Commands that produce a runtime-dependent *dll* app:

Reviewer 2: The wording has .dll as an adjective modifying app, as if .dll is a kind of app. Maybe "... produce a runtime-dependent app as a .dll" would be clearer?

>     Commands that produce a runtime-dependent *dll* app:

Author: Hrm. Maybe it's better to not call out the extension unless talking about the file directly. Instead, could use the terms "platform-specific executable" and "cross-platform binary" to truly differentiate?

>     Commands that produce a runtime-dependent *dll* app:

Reviewer 2: Maybe it isn't necessary to call out the file type at this level.  Maybe your main point is the difference between platform-specific (which is itself a misnomer since it's really runtime-specific) and cross-platform, and you could deal with file types inside the respective sections.  However, you would still have the confusing mismatch between intro and body, where intro is emphasizing cross-platform vs. platform-specific, while the body is emphasizing the difference between self-contained and runtime-dependent.

>     Commands that produce a runtime-dependent *dll* app:

Author: Originally this article was a bit more feature driven (create a FDD, FDE, SCD (framework-dep-deployment, frame-dep-exe, self-contained-exe)) but people are more thinking more about `How do I want a user to start my app: dotnet file.dll or file.exe?` However, if I align with that then there is a lot more repeating info as runtime-dependent spans both dll and exe formats. I think the way it is now lessens how much info is repeated. However, maybe I can redo the table pivot or do a more proper intro after the table pivot.

    | Type                                    | SDK Version    | Command |
    | --------------------------------------- | -------------- | ------- |
    | [runtime-dependent](#runtime-dependent) | 3.1, 3.0, 2.1  | [`dotnet publish`](../tools/dotnet-publish.md) |

    Cross-platform *dll* apps are run with the `dotnet file.dll` command.

## Self-contained

> Self-contained publishing always produces a platform-specific executable. The output publishing folder contains all components of the app, including the .NET Core libraries and target runtime. The app is isolated from other .NET Core apps and doesn't use a locally installed shared runtime. The user of your app isn't required to download and install .NET Core.

Reviewer 2: "publishing" isn't really the thing that is self-contained.  Perhaps "Publishing an app as a self-contained executable always..." Or "When you publish an app as a self-contained executable, the executable is always runtime-specific."

> The executable binary produced is either for the specified target platform. For example, if you have an app named **word_reader**, and you publish a self-contained executable for Windows, a *word_reader.exe* file is created. Publishing for Linux or macOS, a *word_reader* file is created. The target platform and architecture is specified with the `-r <RID>` parameter. For more information about RIDs, see [.NET Core RID Catalog](../rid-catalog.md).

Reviewer 2: Something is missing from the first sentence here or delete "either".

If the app has platform-specific dependencies, such as a NuGet package containing platform-specific dependencies, these are copied to the publish folder along with the app.

> A self-contained app can be created with the [`dotnet publish`](../tools/dotnet-publish.md) command.

Reviewer 2: ```suggestion
A self-contained executable can be created with the [`dotnet publish`](../tools/dotnet-publish.md) command.
```

```dotnet
dotnet publish -r <RID>
```

### Advantages

- **Control .NET Core version**\

> You control which version of .NET Core is deployed with your app. .NET Core can only be upgraded by releasing a new version of your app.

Reviewer 2: The second sentence could be a disadvantage.

- **Platform-specific targeting**\

> Because you have to publish your app for each platform, you know where your app will run. If .NET Core introduces a new platform, users can't run your app on that platform until you release a version that is published that platform. You can test your app for compatibility problems before your users run your app on the new platform.

Reviewer 2: ```suggestion
Because you have to publish your app for each platform, you know where your app will run. If .NET Core introduces a new platform, users can't run your app on that platform until you release a version that is published for that platform. You can test your app for compatibility problems before your users run your app on the new platform.
```

### Disadvantages

> - **Larger deployments**\

Reviewer 2: Given that we're keeping "deployment" in some contexts -- maybe other places it could be used in place of app or executable or dll?

Because your app includes the .NET Core runtime and all of your app dependencies, the hard drive space required is greater than a [runtime-dependent](#runtime-dependent) version.

  > [!TIP]

>   > You can reduce the size of your deployment on Linux systems by approximately 28 MB by using .NET Core [*globalization invariant mode*](https://github.com/dotnet/runtime/blob/master/docs/design/features/globalization-invariant-mode.md). This forces your app to treat all cultures like the [invariant culture](xref:System.Globalization.CultureInfo.InvariantCulture?displayProperty=nameWithType).

Reviewer 2: Why does this not result in a size savings on Windows or macOS deployments?

Author: I'm unsure. Would need someone from the PU to inform us. This not was here in the old version and was specifically written for linux.

> - **More space required across all apps**\

Reviewer 2: This seems like a duplicate of the previous bullet.

Author: yes.. this was here in the original, I didn't like but kept it. I'll remove it though

Each copy of a self-contained app includes its own copy of the .NET Core files, more disk space is consumed.

## Runtime-dependent

Runtime-dependent apps are cross-platform and don't include the .NET Core runtime, but require the user of your app to install the .NET Core runtime.

> A *dll* file is created when you publish your app. For example, if you have an app named **word_reader**, a file named *word_reader.dll* is created. Apps built in this way are run with the `dotnet file.dll` command and can be run on any platform.

Reviewer 2: ```suggestion
A *dll* file is created when you publish your app. For example, if you have an app named **word_reader**, a file named *word_reader.dll* is created. Apps published in this way are run with the `dotnet <filename.dll>` command and can be run on any platform.
```

Along with the *dll* file containing your app, a platform-specific executable can be published if the `-r <RID>` parameter is used. The *dll* is still created and remains cross-platform while the executable isn't. Continuing the example above, if you publish the **word_reader** app and target Windows, a *word_reader.exe* executable is created along with *word_reader.dll*. When targeting Linux or macOS, a *word_reader* executable is created along with *word_reader.dll*. For more information about RIDs, see [.NET Core RID Catalog](../rid-catalog.md).

Starting with .NET Core 3.0 SDK, an executable is always created for your current platform when you omit the `-r` parameter.

If the app uses a NuGet package that has platform-specific implementations, dependencies for all platforms are copied to the publish folder along with the app.

> A runtime-dependent app can be created with the [`dotnet publish`](../tools/dotnet-publish.md) command. With .NET Core 3.0 SDK or higher, both a *dll* and executable file will be created.

Reviewer 2: The second sentence repeats what was said above in the sentence starting with "Starting with .NET Core 3.0 SDK..."

```dotnet
dotnet publish
```

A runtime-dependent app can be created for a specific platform, for example, Linux 64-bit.

```dotnet
dotnet publish -r linux-x64 --self-contained false
```

### Advantages

- **Small deployment**\
Only your app and its dependencies are distributed. The .NET Core runtime and libraries are installed by the user and all apps share the runtime.

- **Cross-platform**\

> Your app and any .NET-based library is easy run on other operating systems. You don't need to define a target platform for your app. For more information about the .NET file format, see [.NET Assembly File Format](../../standard/assembly/file-format.md).

Reviewer 2: ```suggestion
Your app and any .NET-based library runs on other operating systems. You don't need to define a target platform for your app. For more information about the .NET file format, see [.NET Assembly File Format](../../standard/assembly/file-format.md).
```

- **Uses the latest patched runtime**\

> Unless overridden, the app will use the latest runtime (within the targeted major-minor family of .NET Core) installed on the target system. This allows your app to use the latest patched version of the .NET Core runtime.

Reviewer 2: ```suggestion
The app uses the latest runtime (within the targeted major-minor family of .NET Core) installed on the target system. This means your app automatically uses the latest patched version of the .NET Core runtime. This default behavior can be overridden.
```

---
title: Application publishing
description: Learn about the ways to publish a .NET Core application. .NET Core can publish platform-specific or cross-platform apps. You can publish an app as self-contained or as runtime-dependent. Each mode affects how a user runs your app.
ms.date: 01/31/2020
---
# .NET Core application publishing overview

> Applications you create with .NET Core can be published in two different modes, and depending on the mode, affects how a user runs your app. Publishing your app self-contained produces an application that contains the .NET Core runtime and libraries, and your application and its dependencies. Publishing your app runtime-dependent produces an application that contains only your application and its dependencies, and users of your application will need to separately install the .NET Core runtime. Both types of applications include a platform-specific executable by default. Runtime-dependent applications without an executable can be created and these applications are cross-platform. 

Reviewer 2: ```suggestion
Applications you create with .NET Core can be published in two different modes, and the mode affects how a user runs your app.

Publishing your app as *self-contained* produces an application that includes the .NET Core runtime and libraries, and your application and its dependencies. Users of the application can run it on a machine that doesn't have the .NET Core runtime installed.

Publishing your app as *runtime-dependent* produces an application that includes only your application itself and its dependencies. Users of the application have to separately install the .NET Core runtime.

Both publishing modes produce a platform-specific executable by default. Runtime-dependent applications can be created without an executable, and these applications are cross-platform. 
```

> The following table outlines the basic commands used to publish an app as runtime-dependent or self-contained, per SDK:

Reviewer 2: ```suggestion
The following table outlines the commands used to publish an app as runtime-dependent or self-contained, per SDK version:
```

> | Type                                                                  | SDK 3.1 | SDK 3.0 | SDK 2.1 | Command | 

Reviewer 2: Having the version columns in later to earlier order seems counterintuitive to me.  

How about collapsing the 3.x columns into one 3.x column?

Author: I was experimenting with presenting it in the same way as the CLI reference:

![image](https://user-images.githubusercontent.com/2672110/74487425-493b0900-4e74-11ea-8d45-0891c070342c.png)

But I'm not sold on it.

Reviewer 2: In that case the odd order is necessary because the leftmost tab is the default.

Author: Right you are. I swapped them back. It looks better.

| --------------------------------------------------------------------- | --------| ------- | ------- | ------- |
| [runtime-dependent executable](#publish-runtime-dependent)            | ✔️       | ✔️       |         | [`dotnet publish`](../tools/dotnet-publish.md) |

> | [runtime-dependent executable](#publish-runtime-dependent)            | ✔️       | ✔️       |         | [`dotnet publish -r <RID> --self-contained false`](../tools/dotnet-publish.md) |

Reviewer 2: There are two rows with the same text in the type column.  Can we differentiate them, as in 
* ... for the current runtime
* ... for a specified runtime

> | [runtime-dependent cross-platform binary](#publish-runtime-dependent) | ✔️       | ✔️       | ✔️       | [`dotnet publish`](../tools/dotnet-publish.md) |

Reviewer 2: This is confusing to see the exact same command for runtime-dependent executable and runtime-dependent cross-platform binary.  Maybe they could be combined in one line, as "runtime-dependent executable with cross-platform binaries"

Author: I thought about it, but I felt that it's too wordy and would probably wrap. I'll look at using a BR tag and having both commands on the same table row.

> | [self-contained executable](#publish-self-contained)                  | ✔️       | ✔️       | ✔️       | [`dotnet publish -r <RID>`](../tools/dotnet-publish.md) |

Reviewer 2: The acronym RID is used frequently but is never introduced/explained.

> For more information about the `dotnet publish` command, see [.NET Core dotnet publish command](../tools/dotnet-publish.md).

Reviewer 2: ```suggestion
For more information, see [.NET Core dotnet publish command](../tools/dotnet-publish.md).
```

## Produce an executable

> Executables are created per-operating system, per-CPU architecture, and aren't cross-platform. When publishing your app and creating an executable, you can publish the app as [self-contained](#publish-self-contained) or [runtime-dependent](#publish-runtime-dependent). Publishing your app self-contained includes the .NET Core runtime with your app, and users of your app don't have to worry about installing .NET Core before running your app. Apps published runtime-dependent don't include the .NET Core runtime and libraries; only your app and 3rd-party dependencies are included.

Reviewer 2: ```suggestion
Executables aren't cross-platform. They're specific to an operating system and CPU architecture,  When publishing your app and creating an executable, you can publish the app as [self-contained](#publish-self-contained) or [runtime-dependent](#publish-runtime-dependent). Publishing an app as self-contained includes the .NET Core runtime with the app, and users of the app don't have to worry about installing .NET Core before running the app. Apps published as runtime-dependent don't include the .NET Core runtime and libraries; only the app and 3rd-party dependencies are included.
```

The following commands produce an executable:

| Type                                                                  | SDK 3.1 | SDK 3.0 | SDK 2.1 | Command | 
| --------------------------------------------------------------------- | --------| ------- | ------- | ------- |
| [runtime-dependent executable](#publish-runtime-dependent)            | ✔️       | ✔️       |         | [`dotnet publish`](../tools/dotnet-publish.md) |

> | [runtime-dependent executable](#publish-runtime-dependent)            | ✔️       | ✔️       |         | [`dotnet publish -r <RID> --self-contained false`](../tools/dotnet-publish.md) |

Reviewer 2: ditto, it might be helpful to differentiate these rows and collapse the 3.x columns.

| [self-contained executable](#publish-self-contained)                  | ✔️       | ✔️       | ✔️       | [`dotnet publish -r <RID>`](../tools/dotnet-publish.md) |

## Produce a cross-platform binary

Cross-platform binaries are created when you publish your app as [runtime-dependent](#publish-runtime-dependent), in the form of a *dll* file. The *dll* file is named after your project. For example, if you have an app named **word_reader**, a file named *word_reader.dll* is created. Apps published in this way are run with the `dotnet <filename.dll>` command and can be run on any platform.

> Cross-platform binaries can be run on any operating system as long as the targeted .NET Core runtime is already installed. When run, if the targeted .NET Core runtime isn't installed, and your app is configured to roll-forward, your app may still run using a newer runtime. For more information, see [runtime-dependent apps roll forward](../versions/selection.md#framework-dependent-apps-roll-forward).

Reviewer 2: ```suggestion
Cross-platform binaries can be run on any operating system as long as the targeted .NET Core runtime is already installed. If the targeted .NET Core runtime isn't installed, the app may run using a newer runtime if the app is configured to roll-forward. For more information, see [runtime-dependent apps roll forward](../versions/selection.md#framework-dependent-apps-roll-forward).
```

| --------------------------------------------------------------------- | --------| ------- | ------- | ------- |
| [runtime-dependent cross-platform binary](#publish-runtime-dependent) | ✔️       | ✔️       | ✔️       | [`dotnet publish`](../tools/dotnet-publish.md) |

## Publish runtime-dependent

Apps published as runtime-dependent are cross-platform and don't include the .NET Core runtime. The user of your app is required to install the .NET Core runtime.

Publishing an app as runtime-dependent produces a [cross-platform binary](#produce-a-cross-platform-binary) as a *dll* file, and a [platform-specific executable](#produce-an-executable) that targets your current platform. The *dll* is cross-platform while the executable isn't. For example, if you publish an app named **word_reader** and target Windows, a *word_reader.exe* executable is created along with *word_reader.dll*. When targeting Linux or macOS, a *word_reader* executable is created along with *word_reader.dll*. For more information about RIDs, see [.NET Core RID Catalog](../rid-catalog.md).

> [!IMPORTANT]
> .NET Core SDK 2.1 doesn't produce platform-specific executables when you publish an app runtime-dependent.

> The cross-platform binary of your app can be run with the `dotnet <filename.dll>` command, and can be run on any platform. If the app uses a NuGet package that has platform-specific implementations, all platforms dependencies are copied to the publish folder along with the app.

Reviewer 2: ```suggestion
The cross-platform binary of your app can be run with the `dotnet <filename.dll>` command, and can be run on any platform. If the app uses a NuGet package that has platform-specific implementations, all platforms' dependencies are copied to the publish folder along with the app.
```

You can create an executable for a specific platform by passing the `-r <RID> --self-contained false` parameters to the [`dotnet publish`](../tools/dotnet-publish.md) command. When the `-r` parameter is omitted, an executable is created for your current platform. Any NuGet packages that have platform-specific dependencies for the targeted platform are copied to the publish folder.

### Advantages

- **Small deployment**\
Only your app and its dependencies are distributed. The .NET Core runtime and libraries are installed by the user and all apps share the runtime.

- **Cross-platform**\

> Your app and any .NET-based library runs on other operating systems. You don't need to define a target platform for your app. For more information about the .NET file format, see [.NET Assembly File Format](../../standard/assembly/file-format.md).

Reviewer 2: ```suggestion
Your app and any .NET-based library runs on other operating systems. You don't need to define a target platform for your app. For information about the .NET file format, see [.NET Assembly File Format](../../standard/assembly/file-format.md).
```

### Disadvantages

- **Requires pre-installing the runtime**\
Your app can run only if the version of .NET Core your app targets is already installed on the host system. You can configure roll-forward behavior for the app to either require a specific version of .NET Core or allow a newer version of .NET Core. For more information, see [runtime-dependent apps roll forward](../versions/selection.md#framework-dependent-apps-roll-forward).

- **.NET Core may change**\
It's possible for the .NET Core runtime and libraries to be updated on the machine where the app is run. In rare cases, this may change the behavior of your app if you use the .NET Core libraries, which most apps do. You can configure how your app uses newer versions of .NET Core. For more information, see [runtime-dependent apps roll forward](../versions/selection.md#framework-dependent-apps-roll-forward).

The following disadvantage only applies to .NET Core 2.1 SDK.

- **Use the `dotnet` command to start the app**\

> Users must run the `dotnet <filename.dll>` command to start your app. .NET Core 2.1 SDK doesn't produce platform-specific executables for apps published runtime-dependent.

Reviewer 2: ```suggestion
Users must run the `dotnet <filename.dll>` command to start your app. .NET Core 2.1 SDK doesn't produce platform-specific executables for apps published as runtime-dependent.
```

### Advantages

- **Control .NET Core version**\
You control which version of .NET Core is deployed with your app.

- **Platform-specific targeting**\
Because you have to publish your app for each platform, you know where your app will run. If .NET Core introduces a new platform, users can't run your app on that platform until you release a version targeting that platform. You can test your app for compatibility problems before your users run your app on the new platform.

### Disadvantages

- **Larger deployments**\
Because your app includes the .NET Core runtime and all of your app dependencies, the download size and hard drive space required is greater than a [runtime-dependent](#publish-runtime-dependent) version.

  > [!TIP]
  > You can reduce the size of your deployment on Linux systems by approximately 28 MB by using .NET Core [*globalization invariant mode*](https://github.com/dotnet/runtime/blob/master/docs/design/features/globalization-invariant-mode.md). This forces your app to treat all cultures like the [invariant culture](xref:System.Globalization.CultureInfo.InvariantCulture?displayProperty=nameWithType).

> - **Control .NET Core version**\

Reviewer 2: ```suggestion
- **Harder to update the .NET Core version**\
```
